### PR TITLE
[Plugins] Create ArticulatedSystem plugin (originally from SofaGeneralRigid)

### PR DESCRIPTION
**Pull request by fredroy**
ArticulatedSystem is a complex and not really maintained code, designed to create articulated things like articulated bodies, etc.
It was accepted that it should be isolated somewhere and maybe refactored/redesigned later (if someone is motivated).

I took the liberty to retrieve the ArticulatedControllers (which were deprecated/removed some years ago) and put it back into it. Especially because I retrieved the "man-walking" example which is quite cool in my opinion.
But the code was created by DigitalTrainers, I dont know if it could be bothersome license-wise.

**EDIT: Let's vote between**
 - ArticulatedSystem as an external repository which would be referenced in _applications/plugin_  (react with 🚀 )
 - ArticulatedSystem in Sofa.Component because this is really important ( react with ❤️ )

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

**Comment by fredroy** [comment]
[ci-build][with-all-tests]

File: applications/plugins/ArticulatedSystemPlugin/README.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# ArticulatedSystemPlugin
+
+Note: Controllers have created by the company Digital Trainers
```
Comment (alxbilger) [inline applications/plugins/ArticulatedSystemPlugin/README.md]:
```suggestion
Note: Controllers have been created by the company Digital Trainers
```

**Review by alxbilger — APPROVED** [review]
Only a minor change requested

**Comment by guparan** [comment]
I have two questions:
1. Why having "Plugin" in the name? Not sure we have a strict policy on that but it looks like a bad practice to me.
2. Why creating this plugin in SOFA repo and not in an external one?

**Comment by fredroy** [comment]
> I have two questions:
> 
>     1. Why having "Plugin" in the name? Not sure we have a strict policy on that but it looks like a bad practice to me.

Good question, just ArticulatedSystem alone sounded a bit weird to me but I am open to remove the Plugin suffix
 
>     2. Why creating this plugin in SOFA repo and not in an external one?

Another good one 😗, I did not think about it but in my mind the process was ` <in the main codebase> -> "hard" applications/plugins -> "remote" applications/plugins`

**Comment by damienmarchal** [comment]
I have tried to used that one a long time ago to make robotics articulated chain. And I fully agree that some serious improvement can be done. Anyhow, thank for the cleaning @fred as this is a nice first step.

**Comment by fredroy** [comment]
> I have tried to used that one a long time ago to make robotics articulated chain. And I fully agree that some serious improvement can be done. Anyhow, thank for the cleaning @fred as this is a nice first step.

`serious improvement` you are too nice 😅, I guess it would a complete reboot 🏗

**Comment by damienmarchal** [comment]
> > I have tried to used that one a long time ago to make robotics articulated chain. And I fully agree that some serious improvement can be done. Anyhow, thank for the cleaning @fred as this is a nice first step.
> 
> `serious improvement` you are too nice sweat_smile, I guess it would a complete reboot building_construction

:)

**Comment by olivier-goury** [comment]
I jump into this discussion, good job Fred, I like the walking man example, I did not expect it would be possible to do that kind of things in SOFA, it looks cool!

**Comment by fredroy** [comment]
> I jump into this discussion, good job Fred, I like the walking man example, I did not expect it would be possible to do that kind of things in SOFA, it looks cool!

All the credits to the authors of that (ancient) time, and particularly @ChristianDuriez 🤠
This code had been removed from the code base (as nobody seemed to really care), I just merely put it back as I remembered this demo from the good ol' days.

**Comment by adagolodjo** [comment]
Cool @fredroy for resurrecting this code. There is one Rigid left behind, is that normal?

**Comment by fredroy** [comment]
> Cool @fredroy for resurrecting this code. There is one Rigid left behind, is that normal?

No idea, could be a rigid at origin or something like that.
I merely edited the (really old) scene file to be compatible with the new standard, I did not touch anything else.

**Comment by EulalieCoevoet** [comment]
To motivate my vote (an external plugin). I think advertising this code is especially a good idea for us at DEFROST, advertising that we can also simulate articulated system, classic robotics you know. And if it has to be an external plugin to appear on the website (sofa-framework.org/applications/plugins right?)... then I vote for it.  

Edit: I also have a nice example that I could add to the plugin 

image

**Comment by ChristianDuriez** [comment]
> I took the liberty to retrieve the ArticulatedControllers (which were deprecated/removed some years ago) and put it back into it. Especially because I retrieved the "man-walking" example which is quite cool in my opinion. But the code was created by DigitalTrainers, I dont know if it could be bothersome license-wise.

Are you sure ? If it's the case it's only the controller... because the code of the articulatedSystemMapping was done in the SHACRA team...
In any case we had a contract at that time with Digital Trainers on suture that said that the SOFA code developed  in the project was open-source and that Inria keeps the property. 




> 
> * ArticulatedSystem as an external repository which would be referenced in _applications/plugin_  (react with 🚀 )

I vote for a separate plugin to advertise it, as suggested by Eulalie.

One of the thing that should be added to the plugin is the support of Unified Robot Description Format (URDF). 
We should add a loader for it.
This is the new standard format in the community (and compatible with ROS).


Thanks !

Christian

**Comment by fredroy** [comment]
> > I took the liberty to retrieve the ArticulatedControllers (which were deprecated/removed some years ago) and put it back into it. Especially because I retrieved the "man-walking" example which is quite cool in my opinion. But the code was created by DigitalTrainers, I dont know if it could be bothersome license-wise.
> 
> Are you sure ? If it's the case it's only the controller... because the code of the articulatedSystemMapping was done in the SHACRA team... In any case we had a contract at that time with Digital Trainers on suture that said that the SOFA code developed in the project was open-source and that Inria keeps the property.
> 

Yes it is only the controllers. I checked the git log history and it was made by P.J Bensoussan 🧐 on behalf of Digital Trainers.
But good to know that Inria has the property in any case.

**Pull request merged**